# Measuring Income Equality with the Gini Coefficient

**Note**: 

> This exercise has been written out in something called a Jupyter Notebook. We'll discuss Jupyter Notebooks in more detail later in this specialization—they are very a powerful tool for data science communication!—but for the time being, the notebook is just a convenient way for us to write out the exercise. You don't need to *do* anything with the notebook except read its contents—just use write your Python code in a regular `.py` file.

**WARNING:**

> When asked to round your answers to a certain number of decimals, do *not* round any results until you've finished your computations and have your final answer! For example, if you were to calculate the average hourly wage for workers, and you did so by first calculating the average weekly salary of workers and the average hours worked per week, then divided the first number by the second, you should NOT round the average weekly salary of workers or the average hours worked per week. Rounding intermediate results can lead to compounding errors that cause problems for the autograder.

One measure of inequality frequently used by economists, lawyers, social scientists, and policymakers is the *[Gini coefficient](https://en.wikipedia.org/wiki/Gini_coefficient)*. The Gini coefficient takes as input the allocation of some resource — income, wealth, market share, etc. — for all entities in a population and returns a single number between 0 and 1 that summarizes the inequality of the distribution of said resource.

The Gini coefficient takes on a value of 1 when the resource distribution is maximally *unequal* across the entities (e.g., one entity has all of the resource and no one else has any), and a value of 0 when the resource is evenly distributed across all entities. 

In this exercise, we will calculate the Gini coefficient for income inequality across the countries of the world to get a sense of income inequality *across* countries. 


## Exercises

### Exercise 1

To get accustomed to Series, let's explore some data on the wealth of 10 randomly selected countries. Data below presents the GDP per capita for these countries in 2008. 

Use the code below to get started: 

```python
import pandas as pd

gdppercap = pd.Series(
    [34605, 34493, 12393, 44200, 10041, 58138, 4709, 49284, 10109, 42536],
    index=[
        "Bahrain",
        "Belgium",
        "Bulgaria",
        "Ireland",
        "Macedonia",
        "Norway",
        "Paraguay",
        "Singapore",
        "South Africa",
        "Switzerland",
    ],
)
```



In [3]:
import pandas as pd
import numpy as np

gdppercap = pd.Series(
    [34605, 34493, 12393, 44200, 10041, 58138, 4709, 49284, 10109, 42536],
    index=[
        "Bahrain",
        "Belgium",
        "Bulgaria",
        "Ireland",
        "Macedonia",
        "Norway",
        "Paraguay",
        "Singapore",
        "South Africa",
        "Switzerland",
    ],
)

### Exercise 2

Find the mean, median, minimum and maximum values of GDP per capita in this data. Please round all your answers to 1 decimal place.

*Hint:* You *can* do this with several commands, or get all four numbers with a single command.

In [4]:
summary_stats = gdppercap.describe()
print("The summary statistics for the GDP of all countries in this data are:")
summary_stats

The summary statistics for the GDP of all countries in this data are:


count       10.000000
mean     30050.800000
std      19170.095217
min       4709.000000
25%      10680.000000
50%      34549.000000
75%      43784.000000
max      58138.000000
dtype: float64

In [5]:
results = dict()
results["ex2_mean"] = summary_stats.loc["mean"]
results["ex2_median"] = summary_stats.loc["50%"]

In [6]:
results

{'ex2_mean': 30050.8, 'ex2_median': 34549.0}

### Exercise 3

Programmatically, determine which country in our data has the highest income per capita, and which has the lowest income per capita.

(Obviously, this is easier to do by just looking at the data, but that's only because this dataset is very small. With a real dataset, you would need to do it with code, so please write code to accomplish this task.)

Hint: Country names form the index for this Series, so to get country names you'll need to access the index. 


In [4]:
gdppercap = gdppercap.sort_values(ascending=True)
# The ascending=True keyword is not strictly necessary,
# as ascending is the default. But adding it means
# (a) I don't have to trust my memory of the defaults,
# and (b) a reader can easily see what it's dcxoing.
gdppercap.head()

Paraguay         4709
Macedonia       10041
South Africa    10109
Bulgaria        12393
Belgium         34493
dtype: int64

In [5]:
# Richest is last
results["ex3_highest_gdp_percap"] = gdppercap.index[-1]
results["ex3_highest_gdp_percap"]

'Norway'

In [6]:
# Poorest is first
results["ex3_lowest_gdp_percap"] = gdppercap.index[0]
results["ex3_lowest_gdp_percap"]

'Paraguay'

### Exercise 4

Get Python to print out the names of all the countries that have GDP per capita of less than \$20,000. How many countries are in this list?

(As above, try to get pandas to do this _programmatically_ — obviously, this isn't hard to do by just looking at the list in this toy data!)

In [7]:
less_than_20_000 = gdppercap.loc[gdppercap < 20_000].index

In [8]:
results["ex4_lessthan20_000"] = sorted(list(less_than_20_000))

In [9]:
string = ", ".join(results["ex4_lessthan20_000"])
print("The countries in this data with GDP per capitas less than \$2,000 are:")
print(string)

The countries in this data with GDP per capitas less than \$2,000 are:
Bulgaria, Macedonia, Paraguay, South Africa


### Exercise 5 

Get Python to print out the GDP per capita of Switzerland (_programatically_).

In [10]:
gdppercap.loc["Switzerland"]

42536

### Exercise 6

One frequntly used measure of inequality is the Gini Coefficient. The Gini Coefficient takes on a value of 1 when the distribution of some variable is maximally unequal across a population, and a value of 0 when it is evenly distributed. We will calculate the Gini Coefficient for income inequality in our data. 

To visualize the Gini Coefficient, we plot the cumulative share of the population (ordered from poorest to richest) on the x-axis, and cumulative share of income earned by that group on the y-axis. The Gini Coefficient is then defined as $$\frac{A}{A + B}$$, where the areas A and B are labeled below: 

![gini_coefficient](https://upload.wikimedia.org/wikipedia/commons/thumb/5/59/Economics_Gini_coefficient2.svg/800px-Economics_Gini_coefficient2.svg.png)

If income is evenly distributed, then the poorest 20% of a population will also have 20% of the wealth; the poorest 40% will have 40% of the wealth, and so forth, resulting in a perfect 45 degree line. In this situation, there is no area between the 45% line and the actual income distribution, so $A=0$, and the Gini Coefficient is 0. 

If, by contrast, the top 10% of people hold all the wealth in a country, then there will be no wealth for the poorest 90% of people, then wealth will jump up at the far right side of the graph. This will generate a very large gap between the 45% line and actual income for most of the graph, generating a large value for the area $A$, creating a very high Gini Coefficient. 

To illustrate, here are a few different Gini plots. These come from someone studying inequality of participation, so to adapt this to our study of income, just imagine the y-axis plots share of income):

![gini_distributions](https://miro.medium.com/max/595/0*3DTcZnzDwS6A6AtP)

For discrete data, the Gini Coefficient can be calculated with the following formula: 

$$\frac{2 \sum_{i=1}^n i y_i}{n \sum_{i=1}^n y_i} -\frac{n+1}{n}$$

Where $i$ is each country's rank ordering from poorest to richest, and $y_i$ is the income of country $i$.



### Exercise 6

Using this formula, calculate the Gini coefficient for our income data.

Note: if you are not comfortable with the mathematical notation above, you can look it up (just google "summation notation wikipedia" or go to https://en.wikipedia.org/wiki/Summation), or skip this section if you'd like — this is an *ungraded* exercise, so skipping will not impact your ability to progress.

Begin by writing a function to calculate the Gini Coefficient for our data *by looping over the entries in our Series*. In other words, try and embrace the spirit of how you might normally think about interpreting the summation notation written above.

Please round your Gini coefficient to 4 decimal places. Do *not* round any intermediate values you calculate — you should only round your final answer. Rounding intermediate results will result in compounding errors, and may result in an incorrect answer.

**HINT**: Be careful with 0-indexing! Python counts from 0, but mathematical formulas (like $\sum$) start from 1!

**HINT 2**: I'll probably ask you to use this more than once, so please put it in a function.

In [13]:
def gini_w_loops(gdp):
    n = len(gdp)
    gdp = gdp.sort_values(ascending=True)

    # Make top of left side fraction
    top = 0
    for i in range(1, n + 1):
        top = top + gdp.iloc[i - 1] * i

    top = top * 2

    # Make bottom of left side fraction
    bottom = gdp.sum() * len(gdp)

    # make correction
    correction = (n + 1) / n

    # Bring together
    return top / bottom - correction


results["ex6_gini_loop"] = gini_w_loops(gdppercap)
print(f"The cross-country Gini for this data is {results['ex6_gini_loop']:.4f}")

The cross-country Gini for this data is 0.3383


### Exercise 7

Excellent! In data science we generally strive to *not* loop over the entries in our arrays; instead, we aspire to write *vectorized code* that naturally applies a simple operation to each observation.

So now write a new function to calculate the Gini Coefficient that *doesn't* use loops, and instead relies on vectorized code.

Please round your Gini coefficient to 4 decimal places. Do *not* round any intermediate values you calculate — you should only round your final answer. Rounding intermediate results will result in compounding errors, and may result in an incorrect answer.

**HINT:** you will probably have to create some new series/vectors/arrays using a function like `np.arange` or a method like `.rank()`.

In [15]:
def gini_vectorized(gdp):
    n = len(gdp)

    gdp_ranks = gdp.rank(ascending=True)  # returns a series with each entry's rank

    top = 2 * (gdp * gdp_ranks).sum()
    bottom = gdp.sum() * n
    correction = (n + 1) / n

    # Bring together
    return top / bottom - correction


results["ex7_gini_vectorized"] = gini_vectorized(gdppercap)
print(f"The cross-country Gini for this data is {results['ex7_gini_vectorized']:.4f}")

The cross-country Gini for this data is 0.3383


### Exercise 8

The result we just generated offers a snap-shot of inequality for this subset of countries. But what are the dynamics of inequality for these countries?

There is an idea in economics called the "convergence hypothesis", which argues that poorer countries are likely to grow faster, and as a result global inequality is likely to decline. Economists advocating for this hypothesis pointed out that while rich countries had to invent new technologies in order to grow, many poor countries simply had to take advantage of innovations already developed by rich countries. 

To test this hypothesis, let's do a small analysis of the dynamics of income inequality in our sample. Create the following Series in your Python session, which provides the average growth rate of GDP per capita for all the countries in our sample from 2000 to 2018. 

```python
avg_growth = pd.Series(
    [
        -0.29768835,
        0.980299584,
        4.52991925,
        3.686556736,
        2.621416804,
        0.775132075,
        2.015489468,
        3.345793635,
        1.349993318,
        0.982775018,
    ],
    index=[
        "Bahrain",
        "Belgium",
        "Bulgaria",
        "Ireland",
        "Macedonia",
        "Norway",
        "Paraguay",
        "Singapore",
        "South Africa",
        "Switzerland",
    ],
)
```

Using this data on average growth rates in GDP per capita, and assuming growth rates from 2000 to 2018 continue into the future, estimate what our Gini Coefficient may look like in 2025 (remembering that income in our data is from 2008, so we're extrapolating ahead 17 years)?

**Hint:** the formula for compound growth (i.e. value of something growing at a rate of `x` percent for $t$ periods) is:

$$future\_value = current\_value * (1 + \frac{percentage\_growth\_rate}{100}))^t$$

Please round your final Gini coefficient to 4 decimal places.

In [13]:
avg_growth = pd.Series(
    [
        -0.29768835,
        0.980299584,
        4.52991925,
        3.686556736,
        2.621416804,
        0.775132075,
        2.015489468,
        3.345793635,
        1.349993318,
        0.982775018,
    ],
    index=[
        "Bahrain",
        "Belgium",
        "Bulgaria",
        "Ireland",
        "Macedonia",
        "Norway",
        "Paraguay",
        "Singapore",
        "South Africa",
        "Switzerland",
    ],
)
avg_growth

Bahrain        -0.297688
Belgium         0.980300
Bulgaria        4.529919
Ireland         3.686557
Macedonia       2.621417
Norway          0.775132
Paraguay        2.015489
Singapore       3.345794
South Africa    1.349993
Switzerland     0.982775
dtype: float64

In [14]:
gdppercap_2025 = gdppercap * ((avg_growth / 100) + 1) ** 17

In [15]:
gdppercap_2025

Bahrain         32894.836013
Belgium         40714.968220
Bulgaria        26318.849643
Ireland         81790.706907
Macedonia       15589.172902
Norway          66292.970514
Paraguay         6610.779784
Singapore       86235.012247
South Africa    12697.303814
Switzerland     50229.720787
dtype: float64

In [16]:
results["ex8_gini_2025"] = gini_w_loops(gdppercap_2025)

In [17]:
# Or
gini_vectorized(gdppercap_2025)

0.3656264991306195

In [18]:
print(
    f"given these growth rates, the 2025 gini would be {results['ex8_gini_2025']:.3f}"
)

given these growth rates, the 2025 gini would be 0.366


### Exercise 9

Interpret your result -- for this small sample of countries, if growth continued at the rates given, would we see convergence in incomes or not?

[After you're done, you can see a more systematic version of this analysis here!](https://www.cgdev.org/blog/everything-you-know-about-cross-country-convergence-now-wrong)

> Our predicted gini for 2025 is 0.366, while in 2008 the gini was 0.338. So, at least for this sample of countries and growth rates, it does not appear we're seeing convergence; rather, inequality is rising.

In [19]:
assert set(results.keys()) == {
    "ex2_mean",
    "ex2_median",
    "ex3_highest_gdp_percap",
    "ex3_lowest_gdp_percap",
    "ex4_lessthan20_000",
    "ex5_switzerland",
    "ex6_gini_loop",
    "ex7_gini_vectorized",
    "ex8_gini_2025",
}

In [20]:
import json

for k in results.keys():
    if type(results[k]) == np.int64:
        results[k] = int(results[k])

with open("autograder/solutions.json", "w") as outfile:
    json.dump(results, outfile)